session/tlv.go

```diff
@@ -152,3 +154,178 @@ func DeserializeSession(r io.Reader) (*Session, error) {
 
 	return session, nil
 }
+
+// encodeMacaroonRecipe serializes a macaroon recipe so that it can be stored
+// as a tlv value. The serialization will follow the following pattern:
+//
+//	| numPerms | perm1Len | perm1Bytes | perm2Len | perms2Bytes |...
+// 	| numCaveats | caveat1IDLen | caveat1ID | caveat1VerificationIDLen
+//	| caveat1VerificationID | caveat1LocationLen | caveat1Location | ... |
+func encodeMacaroonRecipe(w io.Writer, val interface{},
+	buf *[8]byte) error {
+
+	if v, ok := val.(*MacaroonRecipe); ok {
+		// Encode the macaroon permissions. We do this by first writing
+		// the number of permissions that will be encoded.
+		numPerms := uint64(len(v.Permissions))
+		if err := tlv.WriteVarInt(w, numPerms, buf); err != nil {
+			return err
+		}
+
+		// Now, we iterate over the permissions and for each, we convert
+		// the permission to a string of form `<entity>:<action>` which
+		// we then convert to a byte slice. We first write the length
+		// of the byte slice, followed by the byte slice.
+		for _, perm := range v.Permissions {
+			permBytes := []byte(permissionToString(perm))
+
+			err := tlv.WriteVarInt(w, uint64(len(permBytes)), buf)
+			if err != nil {
+				return err
+			}
+
+			w.Write(permBytes)
+		}
+
+		// Encode macaroon caveats. Again, we first write the number of
+		// caveats that will be encoded.
+		numCaveats := uint64(len(v.Caveats))
+		if err := tlv.WriteVarInt(w, numCaveats, buf); err != nil {
+			return err
+		}
+
+		// Iterate over each caveat and encode them.
+		for _, caveat := range v.Caveats {
+			// Write the length of the caveat ID followed by the ID
+			// itself.
+			err := tlv.WriteVarInt(w, uint64(len(caveat.Id)), buf)
+			if err != nil {
+				return err
+			}
+			w.Write(caveat.Id)
+
+			// Write the length of the caveat verification ID
+			// followed by the verification ID itself.
+			err = tlv.WriteVarInt(
+				w, uint64(len(caveat.VerificationId)), buf,
```

Comment:
Ah I should have mentioned that we're not really going to add third-party caveats any time soon (probably). So the `VerificationId` and `Location` aren't really needed. But let's leave them in now that you've added them, so we're future proof.